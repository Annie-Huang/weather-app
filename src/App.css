
/* SPINNER */
.spinner-wrap {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: 50vh;
}

.spinner {
  width: 100px;
  height: 100px;
  background: #eee;
  border-radius: 50%;
  position: relative;
  margin: 50px;
  display: inline-block;
}

.spinner::after,
.spinner::before {
  content: '';
  display: block;
  width: 200px;
  height: 200px;
  border-radius: 50%;

  position: absolute;
  top: 50%;
  left: 50%;
  margin-top: -100px;
  margin-left: -100px;
  background: #0d6efd;
  animation: pluse 3s linear infinite;
  opacity: 0;
}

.spinner:after {
 animation: pluse 2s linear 2.3s infinite;
}

@keyframes pluse {
  0% {
    transform: scale(0);
    opacity: 1;
  }
  100% {
    transform: scale(1.3);
    opacity: 0;
  }
}

/* WEEKLY FORCAST WIDGET */
.forecast-wrap {
  /*border: 1px solid green;*/
  display: grid;
  grid-template-columns: repeat(auto-fit, 120px);
  gap: 10px;
  justify-content: center;
}
