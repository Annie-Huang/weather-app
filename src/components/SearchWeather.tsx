import React, {ChangeEvent, useState} from 'react';
import './SearchWeather.css';

const SearchWeather = ({city, setCity}) => {
  const [searchCity, setSearchCity] = useState(city);

  const handleSubmitData = () => {
    console.log('annieannie, searchCity=', searchCity);
    setCity(searchCity);
  }

  return (
    <form className='my-3'>
      <label htmlFor="city" className="form-label">Search Weather for your city</label>
      <input type="text" className="form-control" id="city" name="city" value={searchCity} onChange={(e) => setSearchCity(e.target.value)}/>
      city: {city}
      <br />

      Search city: {searchCity}
      <br />

      {/*<div className="form-check mt-2">*/}
      {/*  <input className="form-check-input" type="radio" name="weatherUnit" id="Celsius" checked/>*/}
      {/*    <label className="form-check-label" htmlFor="Celsius">*/}
      {/*      Celsius*/}
      {/*    </label>*/}
      {/*</div>*/}
      {/*<div className="form-check mb-2">*/}
      {/*  <input className="form-check-input" type="radio" name="weatherUnit" id="Fahrenheit" />*/}
      {/*    <label className="form-check-label" htmlFor="Fahrenheit">*/}
      {/*      Fahrenheit*/}
      {/*    </label>*/}
      {/*</div>*/}

      {/*<button type="submit" className="btn btn-primary" onClick={handleSubmitData}>Search</button>*/}
      {/*<button className="btn btn-primary" onClick={handleSubmitData}>Search</button>*/}
      <button className="btn btn-primary" onClick={() => setCity(searchCity)}>Search</button>
    </form>
  );
};

export default SearchWeather;

/*
<input class="form-control form-control-lg" type="text" placeholder=".form-control-lg" aria-label=".form-control-lg example">
<input class="form-control" type="text" placeholder="Default input" aria-label="deafult input example">
<input class="form-control form-control-sm" type="text" placeholder=".form-control-sm" aria-label=".form-control-sm example">
 */

