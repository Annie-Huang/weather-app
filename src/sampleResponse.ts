export const resp = {"data":{"data":[{"moonrise_ts":1598139131,"wind_cdir":"SW","rh":67,"pres":1004.32,"high_temp":11.4,"sunset_ts":1598169172,"ozone":355.646,"moon_phase":0.369673,"wind_gust_spd":11.3134,"snow_depth":0,"clouds":83,"ts":1598104860,"sunrise_ts":1598129622,"app_min_temp":3.7,"wind_spd":5.41524,"pop":75,"wind_cdir_full":"southwest","slp":1013.51,"moon_phase_lunation":0.15,"valid_date":"2020-08-23","app_max_temp":11.4,"vis":22.1113,"dewpt":3.2,"snow":0,"uv":2.02047,"weather":{"icon":"r04d","code":520,"description":"Light shower rain"},"wind_dir":228,"max_dhi":null,"clouds_hi":0,"precip":3.25,"low_temp":5.1,"max_temp":11.7,"moonset_ts":1598190550,"datetime":"2020-08-23","temp":9.3,"min_temp":6.9,"clouds_mid":19,"clouds_low":83},{"moonrise_ts":1598227505,"wind_cdir":"WSW","rh":67,"pres":1012.79,"high_temp":11.4,"sunset_ts":1598255621,"ozone":350.307,"moon_phase":0.488365,"wind_gust_spd":7.42301,"snow_depth":0,"clouds":50,"ts":1598191260,"sunrise_ts":1598215940,"app_min_temp":1.4,"wind_spd":2.4328,"pop":40,"wind_cdir_full":"west-southwest","slp":1022.2,"moon_phase_lunation":0.19,"valid_date":"2020-08-24","app_max_temp":11.4,"vis":22.9915,"dewpt":1.8,"snow":0,"uv":2.9482,"weather":{"icon":"c03d","code":803,"description":"Broken clouds"},"wind_dir":248,"max_dhi":null,"clouds_hi":0,"precip":0.6875,"low_temp":3.9,"max_temp":11.6,"moonset_ts":1598194756,"datetime":"2020-08-24","temp":7.6,"min_temp":4.9,"clouds_mid":0,"clouds_low":50},{"moonrise_ts":1598316033,"wind_cdir":"SSW","rh":72,"pres":1019.9,"high_temp":13,"sunset_ts":1598342071,"ozone":354.454,"moon_phase":0.604832,"wind_gust_spd":5.50676,"snow_depth":0,"clouds":38,"ts":1598277660,"sunrise_ts":1598302258,"app_min_temp":-0.1,"wind_spd":1.70469,"pop":40,"wind_cdir_full":"south-southwest","slp":1029.24,"moon_phase_lunation":0.22,"valid_date":"2020-08-25","app_max_temp":13,"vis":24.1009,"dewpt":3,"snow":0,"uv":3.24863,"weather":{"icon":"c02d","code":802,"description":"Scattered clouds"},"wind_dir":204,"max_dhi":null,"clouds_hi":0,"precip":0.6875,"low_temp":3.9,"max_temp":13.1,"moonset_ts":1598285245,"datetime":"2020-08-25","temp":8,"min_temp":3.7,"clouds_mid":0,"clouds_low":38},{"moonrise_ts":1598404804,"wind_cdir":"WSW","rh":67,"pres":1019.94,"high_temp":14.9,"sunset_ts":1598428520,"ozone":360.821,"moon_phase":0.604832,"wind_gust_spd":7.81116,"snow_depth":0,"clouds":41,"ts":1598364060,"sunrise_ts":1598388575,"app_min_temp":2.4,"wind_spd":1.65104,"pop":0,"wind_cdir_full":"west-southwest","slp":1029.62,"moon_phase_lunation":0.26,"valid_date":"2020-08-26","app_max_temp":14.9,"vis":24.135,"dewpt":3.6,"snow":0,"uv":4.89283,"weather":{"icon":"c03d","code":803,"description":"Broken clouds"},"wind_dir":252,"max_dhi":null,"clouds_hi":13,"precip":0,"low_temp":6.2,"max_temp":14.9,"moonset_ts":1598371645,"datetime":"2020-08-26","temp":9.9,"min_temp":5.8,"clouds_mid":0,"clouds_low":30},{"moonrise_ts":1598493896,"wind_cdir":"NNW","rh":65,"pres":1015.34,"high_temp":17.9,"sunset_ts":1598514970,"ozone":343.142,"moon_phase":0.71266,"wind_gust_spd":10.6074,"snow_depth":0,"clouds":77,"ts":1598450460,"sunrise_ts":1598474891,"app_min_temp":8.5,"wind_spd":3.41726,"pop":25,"wind_cdir_full":"north-northwest","slp":1025.07,"moon_phase_lunation":0.29,"valid_date":"2020-08-27","app_max_temp":17.9,"vis":24.135,"dewpt":5.6,"snow":0,"uv":2.95246,"weather":{"icon":"c04d","code":804,"description":"Overcast clouds"},"wind_dir":339,"max_dhi":null,"clouds_hi":45,"precip":0.3125,"low_temp":8.7,"max_temp":17.9,"moonset_ts":1598461901,"datetime":"2020-08-27","temp":12.3,"min_temp":8.4,"clouds_mid":1,"clouds_low":39},{"moonrise_ts":1598583342,"wind_cdir":"NNW","rh":73,"pres":1013.06,"high_temp":18.8,"sunset_ts":1598601419,"ozone":320.475,"moon_phase":0.806908,"wind_gust_spd":13.3255,"snow_depth":0,"clouds":20,"ts":1598536860,"sunrise_ts":1598561207,"app_min_temp":8.9,"wind_spd":4.21699,"pop":0,"wind_cdir_full":"north-northwest","slp":1022.47,"moon_phase_lunation":0.33,"valid_date":"2020-08-28","app_max_temp":18.1,"vis":24.135,"dewpt":7.7,"snow":0,"uv":5.14076,"weather":{"icon":"c02d","code":801,"description":"Few clouds"},"wind_dir":352,"max_dhi":null,"clouds_hi":0,"precip":0,"low_temp":9.7,"max_temp":18.9,"moonset_ts":1598551803,"datetime":"2020-08-28","temp":12.7,"min_temp":8.9,"clouds_mid":0,"clouds_low":20},{"moonrise_ts":1598673112,"wind_cdir":"W","rh":60,"pres":1007.23,"high_temp":15.3,"sunset_ts":1598687868,"ozone":342.356,"moon_phase":0.88414,"wind_gust_spd":19.2205,"snow_depth":0,"clouds":24,"ts":1598623260,"sunrise_ts":1598647522,"app_min_temp":8.6,"wind_spd":6.49981,"pop":75,"wind_cdir_full":"west","slp":1016.39,"moon_phase_lunation":0.36,"valid_date":"2020-08-29","app_max_temp":14.9,"vis":22.6359,"dewpt":4.3,"snow":0,"uv":3.70046,"weather":{"icon":"c02d","code":802,"description":"Scattered clouds"},"wind_dir":261,"max_dhi":null,"clouds_hi":2,"precip":2.8125,"low_temp":5.5,"max_temp":17.6,"moonset_ts":1598641272,"datetime":"2020-08-29","temp":12,"min_temp":7.5,"clouds_mid":24,"clouds_low":18}],"city_name":"Melbourne","lon":"144.96332","timezone":"Australia/Melbourne","lat":"-37.814","country_code":"AU","state_code":"07"},"status":200,"statusText":"OK","headers":{"content-type":"application/json; charset=utf-8","keep-alive":"timeout=5"},"config":{"url":"http://api.weatherbit.io/v2.0/forecast/daily","method":"get","params":{"key":"4719aae402f54021afa974832544daf9","days":7,"country":"AU","city":"melbourne","units":"M"},"headers":{"Accept":"application/json, text/plain, */*"},"transformRequest":[null],"transformResponse":[null],"timeout":0,"xsrfCookieName":"XSRF-TOKEN","xsrfHeaderName":"X-XSRF-TOKEN","maxContentLength":-1},"request":{}}
