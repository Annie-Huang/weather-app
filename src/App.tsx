import React, {useState} from 'react';
import './App.css';
import useFetchWeather from "./useFetchWeather";
import {IDayWeather} from './model/weather.model';
import DailyForecast from './components/DailyForecast';
import CurrentWeather from './components/CurrentWeather';
import SearchWeather from './components/SearchWeather';
import {WEATHER_UNIT_NAME_MAPPER} from './constant/weather.constant';

const App: React.FC = () => {
  const [city, setCity] = useState('melbourne');
  const [units, setUnits] = useState(WEATHER_UNIT_NAME_MAPPER[0].id); // "M"
  const {loading, days} = useFetchWeather(city, units);

  return (
    <div className='container' style={{maxWidth: '950px'}}>
      <SearchWeather city={city} setCity={setCity} units={units} setUnits={setUnits}/>
      {loading && <div className="spinner-wrap"><div className="spinner" /></div>}
      {!loading &&
        <>
          <CurrentWeather {...days[0]} city={city} units={units}/>
          <div className='forecast-wrap'>
            {days.map((day: IDayWeather, index: number) => <DailyForecast key={index} {...day}/>)}
          </div>
        </>
      }
    </div>
  );
}

export default App;
