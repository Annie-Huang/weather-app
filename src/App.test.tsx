import React from 'react';
import App from './App';
import { mount } from 'enzyme';
import SearchWeather from './components/SearchWeather';
import axios from 'axios';
import { act } from 'react-dom/test-utils';
import { resp } from './sampleResponse.ts';

jest.mock('axios');

it('fetches successfully data from an API', async () => {
  await act(async () => {
    axios.get.mockImplementationOnce(() => Promise.resolve(resp));
    // axios.get.mockImplementationOnce(() => Promise.reject(new Error('Network Error')),);

    const wrapper = mount(<App />);

    wrapper.update();
    // wrapper.forceUpdate();
    // wrapper.instance().forceUpdate();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(SearchWeather)).toHaveLength(1);
  });
});
