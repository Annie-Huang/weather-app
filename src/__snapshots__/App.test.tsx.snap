// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches successfully data from an API 1`] = `
<App>
  <div
    className="container"
    style={
      Object {
        "maxWidth": "950px",
      }
    }
  >
    <SearchWeather
      city="melbourne"
      setCity={[Function]}
      setUnits={[Function]}
      units="M"
    >
      <form
        className="my-3 d-flex flex-column justify-content-center"
      >
        <label
          className="form-label"
          htmlFor="city"
        >
          Search Weather for your city
        </label>
        <input
          aria-label="search city"
          className="form-control mb-2"
          id="city"
          name="city"
          onChange={[Function]}
          type="text"
          value="melbourne"
        />
        <button
          className="btn btn-primary"
          onClick={[Function]}
        >
          Search
        </button>
        <div
          className="d-flex justify-content-center"
        >
          <div
            className="form-check mt-2 mr-4"
            key="0"
          >
            <input
              checked={true}
              className="form-check-input"
              id="Celsius"
              name="weather-unit"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="form-check-label"
              htmlFor="Celsius"
            >
              Celsius
            </label>
          </div>
          <div
            className="form-check mt-2 mr-4"
            key="1"
          >
            <input
              checked={false}
              className="form-check-input"
              id="Fahrenheit"
              name="weather-unit"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="form-check-label"
              htmlFor="Fahrenheit"
            >
              Fahrenheit
            </label>
          </div>
        </div>
      </form>
    </SearchWeather>
    <div
      className="spinner-wrap"
    >
      <div
        className="spinner"
      />
    </div>
  </div>
</App>
`;
